README.md
# spectrum-generator

[![NPM version][npm-image]][npm-url]
[![build status][ci-image]][ci-url]
[![npm download][download-image]][download-url]

Generates a spectrum from discrete peaks. The shape of the peaks can be customised.

In order to increase the speed a `shape` is first generated and then the peaks in the final
spectrum are resulting from sampling the `shape`. A `shape` will therefore be generated with
much more points (typically fwhm:1000).

## Installation

`$ npm i spectrum-generator`

## Usage

The spectrum generator takes an array of discrete peaks (value and intensity)
and generates peaks with a gaussian distribution.

### generateSpectrum

```js
import { generateSpectrum } from 'spectrum-generator';

const peaks = [
  [4, 10],
  [20, 30],
  [236, 1],
  [569, 76],
];
const spectrum = generateSpectrum(peaks, {
  from: 0, // default value: 0
  to: 1000, // default value: 1000
  nbPoints: 10001, // default value: 10001
});
```

Example to generate a high resolution spectrum using lorentzian peaks shape

```js
import { generateSpectrum } from 'spectrum-generator';

const peaks = [
  [1, 10],
  [2, 30],
  [3, 1],
  [4, 76],
];

/*
Peaks can also be introduced as:
- const peaks = [{x:1,y:10},{x:2,y:30},{x:3,y:1},{x:4,y:76}]
- const peaks = {x:[1,2,3,4], y:[10,30,1,76]}
*/

const spectrum = generateSpectrum(peaks, {
  nbPoints: 1001,
  from: 0,
  to: 10,
  shape: {
    kind: 'lorentzian',
    options: {
      fwhm: 1000,
      length: 10001,
    },
  },
});
```

### class SpectrumGenerator

```js
import { SpectrumGenerator } from 'spectrum-generator';

const generator = new SpectrumGenerator();
generator.addPeak([5, 20]);
generator.addPeak({ x: 5, y: 20 }); // we may either add an array of 2 elements or an object with x,y values
generator.addPeak([30, 56]);
generator.addPeaks([
  [40, 12], // it can also be an array of objects with x,y properties
  [10, 1],
]);
const spectrum = generator.getSpectrum();

generator.reset();

generator.addPeak([10, 50], { // customize peaks shape
  width: 0.1,
  shape: {
    kind: 'lorentzian',
    options: {
      fwhm: 1000,
      length: 10001,
    },
  }
});

generator.addPeak([10, 50], { // customize peaks shape
  width: 0.1,
  shape: {
    kind: 'gaussian',
    options: {
      fwhm: 1000,
      length: 10001,
    },
  }
});
const otherSpectrum = generator.getSpectrum();
```

## [API Documentation](https://cheminfo.github.io/spectrum-generator/)

## License

[MIT](./LICENSE)

[npm-image]: https://img.shields.io/npm/v/spectrum-generator.svg
[npm-url]: https://www.npmjs.com/package/spectrum-generator
[ci-image]: https://github.com/cheminfo/spectrum-generator/workflows/Node.js%20CI/badge.svg?branch=master
[ci-url]: https://github.com/cheminfo/spectrum-generator/actions?query=workflow%3A%22Node.js+CI%22
[download-image]: https://img.shields.io/npm/dm/spectrum-generator.svg
[download-url]: https://www.npmjs.com/package/spectrum-generator


History.md
## [4.1.0](https://github.com/cheminfo/spectrum-generator/compare/v4.0.2...v4.1.0) (2020-10-12)


### Features

* addPeaks with {x:[], y:[]} ([#11](https://github.com/cheminfo/spectrum-generator/issues/11)) ([3ac6f5e](https://github.com/cheminfo/spectrum-generator/commit/3ac6f5eb4f754a632fb1acd24890a1d08921de9a))

## [4.0.2](https://github.com/cheminfo/spectrum-generator/compare/v4.0.1...v4.0.2) (2020-04-18)


### Bug Fixes

* one more edge bug that leads to y=0 ([8bbca0f](https://github.com/cheminfo/spectrum-generator/commit/8bbca0fe8659290e15d413b6a8f2b28d071dd877))



## [4.0.1](https://github.com/cheminfo/spectrum-generator/compare/v4.0.0...v4.0.1) (2020-04-18)


### Bug Fixes

* example SpectrumGenerator ([cf49ce1](https://github.com/cheminfo/spectrum-generator/commit/cf49ce1db859e542bb6324006ed338546b4d390f))
* examples ([4a8289b](https://github.com/cheminfo/spectrum-generator/commit/4a8289bafc894d27158c7c4bbadc0d74f1c84206))
* rouding problem in spectrum generation yield to zero values ([fcb1521](https://github.com/cheminfo/spectrum-generator/commit/fcb1521014441290b345d758412c9922f714a156))



# [4.0.0](https://github.com/cheminfo/spectrum-generator/compare/v3.2.2...v4.0.0) (2020-03-18)



## [3.2.2](https://github.com/cheminfo/spectrum-generator/compare/v3.2.1...v3.2.2) (2020-03-13)


### Bug Fixes

* remove cheminfo-tools from dependencies ([5c17e1c](https://github.com/cheminfo/spectrum-generator/commit/5c17e1c2b48f50309fc0f3f23fd596221e33d86a))



## [3.2.1](https://github.com/cheminfo/spectrum-generator/compare/v3.2.0...v3.2.1) (2020-03-07)


### Bug Fixes

* ensure peak symmetry ([128def1](https://github.com/cheminfo/spectrum-generator/commit/128def19cf1932f1fd853bb6b983a91559022666))



# [3.2.0](https://github.com/cheminfo/spectrum-generator/compare/v3.1.3...v3.2.0) (2020-03-02)


### Features

* allow to specify the shape ([abb6a7d](https://github.com/cheminfo/spectrum-generator/commit/abb6a7df2cdbe198ae32162bb98765b2f66a0ae3))



# [3.1.0](https://github.com/cheminfo/spectrum-generator/compare/v3.0.1...v3.1.0) (2019-01-10)



<a name="3.0.1"></a>
## [3.0.1](https://github.com/cheminfo/spectrum-generator/compare/v3.0.0...v3.0.1) (2018-10-23)


### Bug Fixes

* **addPeak:** fix peak generation with non-integer middlepoint index ([#3](https://github.com/cheminfo/spectrum-generator/issues/3)) ([baf1a6e](https://github.com/cheminfo/spectrum-generator/commit/baf1a6e))



<a name="3.0.0"></a>
# [3.0.0](https://github.com/cheminfo/spectrum-generator/compare/v2.0.1...v3.0.0) (2018-10-23)



<a name="2.0.1"></a>
## [2.0.1](https://github.com/cheminfo/spectrum-generator/compare/v2.0.0...v2.0.1) (2018-08-03)



<a name="2.0.0"></a>
# [2.0.0](https://github.com/cheminfo/spectrum-generator/compare/v1.1.0...v2.0.0) (2018-07-28)



<a name="1.1.0"></a>
# [1.1.0](https://github.com/cheminfo/spectrum-generator/compare/v1.0.1...v1.1.0) (2018-03-12)



<a name="1.0.1"></a>
## [1.0.1](https://github.com/cheminfo/spectrum-generator/compare/v1.0.0...v1.0.1) (2018-03-09)


### Bug Fixes

* fix bug when start not 0 and pointsPerUnit not 1 ([a41d3db](https://github.com/cheminfo/spectrum-generator/commit/a41d3db))



<a name="1.0.0"></a>
# 1.0.0 (2017-11-09)


### Features

* implement spectrum generator ([6199252](https://github.com/cheminfo/spectrum-generator/commit/6199252))





src/index.js
import normed from 'ml-array-normed';
import { getShape } from 'ml-peak-shape-generator';

import addBaseline from './util/addBaseline.js';
import addNoise from './util/addNoise.js';

export class SpectrumGenerator {
  constructor(options = {}) {
    options = Object.assign(
      {},
      {
        from: 0,
        to: 1000,
        nbPoints: 10001,
        peakWidthFct: () => 5,
        shape: {
          kind: 'gaussian',
          options: {
            fwhm: 1000,
            length: 5001,
          },
        },
      },
      options,
    );
    this.from = options.from;
    this.to = options.to;
    this.nbPoints = options.nbPoints;
    this.interval = (this.to - this.from) / (this.nbPoints - 1);
    this.peakWidthFct = options.peakWidthFct;
    this.maxPeakHeight = Number.MIN_SAFE_INTEGER;
    this.shape = createShape(options.shape.kind, options.shape.options);

    assertNumber(this.from, 'from');
    assertNumber(this.to, 'to');
    assertInteger(this.nbPoints, 'nbPoints');

    if (this.to <= this.from) {
      throw new RangeError('to option must be larger than from');
    }

    if (typeof this.peakWidthFct !== 'function') {
      throw new TypeError('peakWidthFct option must be a function');
    }

    this.reset();
  }

  addPeaks(peaks, options) {
    if (
      !Array.isArray(peaks) &&
      (typeof peaks !== 'object' ||
        peaks.x === undefined ||
        peaks.y === undefined ||
        !Array.isArray(peaks.x) ||
        !Array.isArray(peaks.y) ||
        peaks.x.length !== peaks.y.length)
    ) {
      throw new TypeError(
        'peaks must be an array or an object containing x[] and y[]',
      );
    }
    if (Array.isArray(peaks)) {
      for (const peak of peaks) {
        this.addPeak(peak, options);
      }
    } else {
      for (let i = 0; i < peaks.x.length; i++) {
        this.addPeak([peaks.x[i], peaks.y[i]], options);
      }
    }

    return this;
  }

  addPeak(peak, options = {}) {
    if (
      typeof peak !== 'object' ||
      (peak.length !== 2 && (peak.x === undefined || peak.y === undefined))
    ) {
      throw new Error(
        'peak must be an array with two values or an object with {x,y}',
      );
    }
    let xPosition;
    let intensity;
    if (Array.isArray(peak)) {
      [xPosition, intensity] = peak;
    } else {
      xPosition = peak.x;
      intensity = peak.y;
    }

    if (intensity > this.maxPeakHeight) this.maxPeakHeight = intensity;

    let {
      width = this.peakWidthFct(xPosition),
      widthLeft,
      widthRight,
      shape: shapeOptions,
    } = options;

    const shape = shapeOptions
      ? createShape(shapeOptions.kind, shapeOptions.options)
      : this.shape;

    if (!widthLeft) widthLeft = width;
    if (!widthRight) widthRight = width;

    const firstValue = xPosition - (widthLeft / 2) * this.shape.factor;
    const lastValue = xPosition + (widthRight / 2) * this.shape.factor;

    const firstPoint = Math.max(
      0,
      Math.floor((firstValue - this.from) / this.interval),
    );
    const lastPoint = Math.min(
      this.nbPoints - 1,
      Math.ceil((lastValue - this.from) / this.interval),
    );
    const middlePoint = Math.round((xPosition - this.from) / this.interval);
    // PEAK SHAPE MAY BE ASYMMETRC (widthLeft and widthRight) !
    // we calculate the left part of the shape
    for (let index = firstPoint; index < middlePoint; index++) {
      let ratio = ((xPosition - this.data.x[index]) / widthLeft) * 2;
      let shapeIndex = Math.round(shape.halfLength - (ratio * shape.fwhm) / 2);
      if (shapeIndex >= 0 && shapeIndex < shape.data.length) {
        this.data.y[index] += shape.data[shapeIndex] * intensity;
      }
    }
    // we calculate the right part of the gaussian
    for (let index = middlePoint; index <= lastPoint; index++) {
      let ratio = ((this.data.x[index] - xPosition) / widthRight) * 2;

      let shapeIndex = Math.round(shape.halfLength - (ratio * shape.fwhm) / 2);
      if (shapeIndex >= 0 && shapeIndex <= shape.data.length) {
        this.data.y[index] += shape.data[shapeIndex] * intensity;
      }
    }

    return this;
  }

  addBaseline(baselineFct) {
    addBaseline(this.data, baselineFct);
    return this;
  }

  addNoise(percent, options) {
    addNoise(this.data, percent, options);
    return this;
  }

  getSpectrum(options = {}) {
    if (typeof options === 'boolean') {
      options = { copy: options };
    }
    const { copy = true, threshold = 0 } = options;
    if (threshold) {
      let minPeakHeight = this.maxPeakHeight * threshold;
      let x = [];
      let y = [];
      for (let i = 0; i < this.data.x.length; i++) {
        if (this.data.y[i] >= minPeakHeight) {
          x.push(this.data.x[i]);
          y.push(this.data.y[i]);
        }
      }
      return { x, y };
    }
    if (copy) {
      return {
        x: this.data.x.slice(),
        y: this.data.y.slice(),
      };
    } else {
      return this.data;
    }
  }

  reset() {
    const spectrum = (this.data = {
      x: new Float64Array(this.nbPoints),
      y: new Float64Array(this.nbPoints),
    });

    for (let i = 0; i < this.nbPoints; i++) {
      spectrum.x[i] = this.from + i * this.interval;
    }

    return this;
  }
}

function assertInteger(value, name) {
  if (!Number.isInteger(value)) {
    throw new TypeError(`${name} option must be an integer`);
  }
}

function assertNumber(value, name) {
  if (!Number.isFinite(value)) {
    throw new TypeError(`${name} option must be a number`);
  }
}

export function generateSpectrum(peaks, options = {}) {
  const generator = new SpectrumGenerator(options);
  generator.addPeaks(peaks);
  if (options.baseline) generator.addBaseline(options.baseline);
  if (options.noise) generator.addNoise(options.noise.percent, options.noise);
  return generator.getSpectrum({
    threshold: options.threshold,
  });
}

function createShape(kind, options) {
  let shape = {};
  let newShape = getShape(kind, options);
  shape.data = normed(newShape.data, {
    algorithm: 'max',
  });
  shape.fwhm = newShape.fwhm;
  shape.factor = (newShape.data.length - 1) / newShape.fwhm;
  shape.length = newShape.data.length;
  shape.halfLength = Math.floor(newShape.data.length / 2);
  return shape;
}
